File: app/src/main/java/com/univreview/view/presenter/UploadReviewPresenter.java
package com.univreview.view.presenter;

import android.content.Context;

import com.univreview.App;
import com.univreview.fragment.MypageFragment;
import com.univreview.log.Logger;
import com.univreview.model.Review;
import com.univreview.network.Retro;
import com.univreview.util.ErrorUtils;
import com.univreview.util.Util;
import com.univreview.view.contract.UploadReviewContract;

import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by DavidHa on 2017. 7. 20..
 */

public class UploadReviewPresenter implements UploadReviewContract {
    private UploadReviewContract.View view;
    private Review review;
    private Context context;


    @Override
    public void registerReview() {
        review.userId = App.userId;
        review.subjectDetail.subject.name = view.getSubjectName();
        review.subjectDetail.professor.name = view.getProfessorName();
        if (review.getAlertMessage() == null) {
            callPostSimpleReviewApi(review);
            view.showProgress();
        } else {
            Util.simpleMessageDialog(context, review.getAlertMessage());
        }
    }

    private void callPostSimpleReviewApi(Review review){
        Logger.v("post review: " + review);
        Logger.v("prof id: " + review.professorId);
        Logger.v("subj id: " + review.subjectId);
        Retro.instance.reviewService().postSimpleReview(App.setAuthHeader(App.userToken), review)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doAfterTerminate(() -> view.dismissProgress())
                .subscribe(result -> response(result.review), ErrorUtils::parseError);
    }

    @Override
    public void checkReviewExist() {
        Retro.instance.reviewService().getReviewExist(App.setAuthHeader(App.userToken), review.subjectId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(result -> {
                    view.setReviewExist(result.exist);
                    if (result.exist) {
                        view.showAlertDialog();
                    }
                }, ErrorUtils::parseError);
    }

    private void response(Review review) {
        Logger.v("response review: " + review);
        MypageFragment.isRefresh = true;
        this.review.id = review.id;
        view.showRecommendRvDialog();
    }

    @Override
    public Review getReview() {
        return review;
    }

    @Override
    public void setContext(Context context) {
        this.context = context;
    }

    @Override
    public void attachView(View view) {
        this.view = view;
        this.review = new Review();
    }

    @Override
    public void detachView() {
        view = null;
        review = null;
    }
}

File: app/src/main/java/com/univreview/view/presenter/UploadReviewPresenter.kt
package com.univreview.view.presenter

import android.content.Intent
import com.univreview.App
import com.univreview.log.Logger
import com.univreview.model.enumeration.ReviewSearchType
import com.univreview.model.Review
import com.univreview.network.Retro
import com.univreview.util.ErrorUtils
import com.univreview.view.contract.UploadReviewContract
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

/**
 * Created by DavidHa on 2017. 8. 6..
 */
class UploadReviewPresenter(val review: Review = Review()) : UploadReviewContract {
    lateinit var view: UploadReviewContract.View

    override fun registerReview() {
        if (review.getAlertMessage() == null) {
            callPostSimpleReviewApi(review)
            view.showProgress()
        } else {
            view.showSimpleMsgDialog(review.getAlertMessage().toString())
        }
    }

    private fun callPostSimpleReviewApi(review: Review) {
        Logger.v("post review: " + review)
        Logger.v("course id: " + review.courseId)
        Logger.v("subj id: " + review.subjectId)
        Retro.instance.reviewService.callPostReview(App.setHeader(), review)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doAfterTerminate { view.dismissProgress() }
                .subscribe({ response(it.data) }, { ErrorUtils.parseError(it) })
    }

    override fun checkReviewExist() {
        Retro.instance.reviewService.callCheckReviewForCourseId(App.setHeader(), review.courseId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it.isExist) {
                        view.showAlertDialog()
                        review.courseId = 0
                    }
                }, { ErrorUtils.parseError(it) })
    }

    private fun response(review: Review) {
        Logger.v("response review: " + review)
        view.showRecommendRvDialog(review)
    }

    override fun onActivityResult(data: Intent) {
        val id = data.getLongExtra("id", 0)
        val name = data.getStringExtra("name")
        val type = data.getSerializableExtra("type") as ReviewSearchType
        if (ReviewSearchType.SUBJECT_WITH_RESULT == type) {
            view.setSubjectTxt(name)
            view.setProfessorTxt(null)
            review.subjectId = id
            review.courseId = 0
        } else if (ReviewSearchType.PROF_FROM_SUBJ == type) {
            view.setProfessorTxt(name)
            review.courseId = id
            checkReviewExist()
        }
    }

}